### PR TITLE
core: Simplify successor terminator conditions

### PR DESCRIPTION
**Pull request by compor**
This PR simplifies and fixes the successor conditions for terminator verification and improves exception messages by adding the operation name.

This is a spinoff from #1049 according to [these steps](https://github.com/xdslproject/xdsl/pull/1049#issuecomment-1594290198).

### CODE REVIEW

**Review by superlopuh — APPROVED** [review]

**Review by AntonLydike — APPROVED** [review]

**Review by webmiche — APPROVED** [review]

File: xdsl/ir/core.py [hunk 1]
```diff
@@ -769,30 +769,38 @@ def verify(self, verify_nested_ops: bool = True) -> None:
...
+                    )
 
+        if parent_block is not None and parent_region is not None:
             # TODO single-block regions dealt when the NoTerminator trait is
             # implemented (https://github.com/xdslproject/xdsl/issues/1093)
             if len(parent_region.blocks) > 1:
-                if parent_block.last_op is self and not self.has_trait(IsTerminator):
+                if (
+                    parent_block.last_op is self
+                    and len(self.successors) == 0
```
Comment (math-fehr) [inline xdsl/ir/core.py:798]:
```suggestion
```
Even if an operation has successor, it should be a terminator?

File: tests/test_ir.py [hunk 2]
```diff
@@ -170,16 +168,13 @@ def test_op_clone_with_regions():
...
-    Utility operation that requires a successor and has the IsTerminator
-    trait.
+    Utility operation that requires a successor.
     """
 
     name = "test.successor_op"
 
     successor: Successor = successor_def()
 
-    traits = frozenset([IsTerminator()])
```
Comment (math-fehr) [inline tests/test_ir.py:181]: This looks suspicious to me.
If an operation has successor, it needs the terminator trait.

Comment (math-fehr) [reply tests/test_ir.py:181]: test_block_not_branching_to_another_region should fail because of that (also because the first block is empty)

**Review by math-fehr — COMMENTED** [review]
Looks good to me, but I just don't understand why operations with successors do not need the terminator trait?
AFAIK, even if an operation has successors, it should have the terminator trait to end the block. Essentially, having successors but not being a terminator should always be a bug.

File: xdsl/ir/core.py (hunk 1 above)
Comment (compor) [reply xdsl/ir/core.py:798]: Yeah, that's a good catch, I re-read the MLIR `Verifier.cpp` wrong on this.

**Comment by compor** [comment]
> Looks good to me, but I just don't understand why operations with successors do not need the terminator trait? AFAIK, even if an operation has successors, it should have the terminator trait to end the block. Essentially, having successors but not being a terminator should always be a bug.

There was a test case missing that exposed this, it's been added.

**Review by math-fehr — APPROVED** [review]
Perfect! Thanks!

**Pull request merged**